pkg/minikube/storageclass/storageclass_test.go

```diff
@@ -0,0 +1,269 @@
+clusters:
+- cluster:
+    server: https://example.com:443
+  name: minikube
+contexts:
+- context:
+    cluster: minikube
+    user: minikube
+  name: minikube
+current-context: minikube
+kind: Config
+preferences: {}
+users:
+- name: minikube
+`
+
+func mockGetStoragev1() (storagev1.StorageV1Interface, error) {
+	client := fake.Clientset{Fake: testing_client.Fake{}}
+	s := client.StorageV1()
+	return mockStorageV1Interface{s}, nil
+}
+
+func (mockStorageV1Interface) StorageClasses() storagev1.StorageClassInterface {
+	return mockStorageClassInterface{}
+}
+
+type mockStorageV1Interface struct {
+	storagev1.StorageV1Interface
+}
+
+type mockStorageClassInterface struct {
+	storagev1.StorageClassInterface
+}
+
+func (mockStorageClassInterface) Get(name string, options metav1.GetOptions) (*v1.StorageClass, error) {
+	if name == "no-such-class" {
+		return nil, fmt.Errorf("mocked error. No such class")
+	}
+	sc := v1.StorageClass{Provisioner: name}
+	return &sc, nil
+}
+
+var mockStorageProvisioner string
+
+func (mockStorageClassInterface) List(opts metav1.ListOptions) (*v1.StorageClassList, error) {
+	scl := v1.StorageClassList{}
+	scl.Items = append(scl.Items, v1.StorageClass{Provisioner: mockStorageProvisioner})
+	return &scl, nil
+}
+
+func (mockStorageClassInterface) Update(sc *v1.StorageClass) (*v1.StorageClass, error) {
+	if strings.HasPrefix(sc.Provisioner, "bad") {
+		return nil, fmt.Errorf("bad provisioner")
+	}
+	s := v1.StorageClass{}
+	return &s, nil
+}
+
+func TestSetDefaultStorageClassNoFake(t *testing.T) {
+	originalEnv := os.Getenv("KUBECONFIG")
+	defer func() {
+		err := os.Setenv("KUBECONFIG", originalEnv)
+		if err != nil {
+			t.Fatalf("Error reverting env KUBECONFIG to its original value. Got err (%s)", err)
+		}
+	}()
+	var tests = []struct {
+		description    string
+		kubeconfigPath string
+		config         string
+		class          string
+		err            bool
+	}{
+		{
+			description:    "list storage class error",
+			kubeconfigPath: "/tmp/kube_config",
+			config:         mockK8sConfig,
+			class:          "standard",
+			err:            true,
+		},
+		{
+			description:    "broken config",
+			kubeconfigPath: "/tmp/kube_config",
+			config:         "this**is&&not: yaml::valid: file",
+			class:          "standard",
+			err:            true,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.description, func(t *testing.T) {
+			if err := setK8SConfig(test.config, test.kubeconfigPath); err != nil {
+				t.Fatalf(err.Error())
+			}
+			defer func() { os.Remove(test.kubeconfigPath) }()
+
+			err := SetDefaultStorageClass(test.class)
+			if err != nil && !test.err {
+				t.Fatalf("Unexpected err: %v for test: %v", err, test.description)
+			}
+			if err == nil && test.err {
+				t.Fatalf("Expected err for test: %v", test.description)
+			}
+		})
+	}
+}
+
+func TestSetDefaultStorageClassFakeStorage(t *testing.T) {
+	getStoragev1 = mockGetStoragev1
+	defer func() { getStoragev1 = orgGetStoragev1 }()
```

Comment:
Ah! That's why you created the package variable. For me having to do this here shows a design flaw. The storage is a dependency, it should be passed on the API, instead of being hidden in the internals. I prefer we change the API to:

```
func GetStoragev1() (storagev1.StorageV1Interface, error) {...}
```

and pass it to the methods that will use it as a dependency:
```
func DisableDefaultStorageClass(storage storagev1.StorageV1Interface, class string) error {}
func SetDefaultStorageClass(storage storagev1.StorageV1Interface, name string) error {}
```

And remove this package variable, and also this magic with defers to tests, with this been injected in the API you will be able to pass the custom mock to test it.

Reply:
Refactored. Seems much cleaner now. Are you fine with similar API changes for the future?

Reply:
Sure! But remember that smaller PRs with fewer changes are easier to review. So depending on the change is better to break into different PRs.